### DISCUSSION TITLE
Major Release 3

### DISCUSSION THREAD

**Discussion by maennchen** [maennchen/ZipStream-PHP#237] [Show and tell]
**EDIT: v3 has been released. If you face any issues, please open issues instead of commenting here.**

:tada: We've just released the new major version 3 (beta).

## News

* Requires PHP >= 8.1 (Version 2 will be supported until PHP 8.0 is no longer maintained)
* Extensively uses new PHP Features
  - Named Arguments
  - Enums
 * The code was rewritten to simplify maintenance. Hopefully, that will lead to more people being able to contribute to the project.
 * Supports to create of empty directories

## Upgrade Instructions

https://github.com/maennchen/ZipStream-PHP/tree/3.0.0-beta.2#upgrade-to-version-300

## Feedback

Plese try out the beta and let us know if you are facing issues / suggested improvements in the discussion.

Please be aware that this is a pre-release. Do not use it in production yet and the final release could still contain breaking changes based on the received feedback.

**Comment by NicolasCARPi** [comment]
Hello @maennchen.

It's great that you found the time to do a refactor :tada: 

I finally was able to take the time to test this new release, and as far as I can tell, it seems to work fine!

Basically, my diff looks like this:

~~~diff
-        $opt = new ArchiveOptions();
-        // crucial option for a stream output
-        $opt->setZeroHeader(true);
-        return new ZipStream(null, $opt);
+        return new ZipStream(sendHttpHeaders:false);
~~~

Using named arguments is a good idea.

I suggest you go ahead and publish the final 3.0.0.

Best,
~Nico

**Reply by NicolasCARPi** [reply]
So in fact I have an issue. The `ZipStream` class being `final`, it cannot be mocked. It seems there could be [workarounds](https://tomasvotruba.com/blog/2019/03/28/how-to-mock-final-classes-in-phpunit/), but I believe there would still be an issue with phpstan and static analysis.

I think the most easy and straightforward thing to do would be to simple remove the `final` keyword from the class. It would probably save some time to other devs when they upgrade, and I don't believe there's much downside TBH. What do you think @maennchen ?

**Reply by maennchen** [reply]
@NicolasCARPi Agreed.

I also want to finalize #241 before I go ahead and tag stable since I still might need to do some changes.

**Reply by maennchen** [reply]
And thanks for sponsoring me, that is much appreciated! ❤️

**Reply by NicolasCARPi** [reply]
Ah, I realize the `main` branch doesn't have `final` already. All good.

**Reply by NicolasCARPi** [reply]
> I also want to finalize https://github.com/maennchen/ZipStream-PHP/pull/241 before I go ahead

I would propose to go ahead right now and leave that non-critical feature for 3.1. It doesn't look like it'll break everything anyway. I'm suggesting this to avoid 3.0 going into limbo because of this artificial blocker ;)

**Reply by maennchen** [reply]
@NicolasCARPi Might be a good idea as well...
I'll think about it :)

**Reply by maennchen** [reply]
@NicolasCARPi Released :+1:

**Reply by NicolasCARPi** [reply]
way ahead of you: https://github.com/elabftw/elabftw/pull/4164 ;)
